ENH: HiGHs re-integration

HaoZeke opened the pull request:
Clean-room implementation of #21226 and therefore #19255. Needs a companion PR merged at `SciPy/HiGHs`.  Should be easier to review, since the changes are minimal. CI tests ~~on my fork worked fine, with one major exception (in a comment to this)~~ should be green now too.

Upgrades HiGHS from 1.4.0 to 1.8.0

- [x] Retarget to the SciPy fork (after the companion PR is merged)

## Companion PRs
- [x] SciPy fork of HiGHS --> https://github.com/scipy/HiGHS/pull/69
- [x] HiGHS upstream --> https://github.com/ERGO-Code/HiGHS/pull/2005
- [x] Build warning suppression (upstream) --> https://github.com/ERGO-Code/HiGHS/pull/2023/files

rgommers committed:
BLD: move the `optimize` build steps earlier into the build sequence

This makes the build more balanced, since HiGHS takes a long time
to build and finishes last otherwise.

rgommers committed:
BENCH: disable very slow benchmark in `optimize_milp.py`

This was no slower after the update to HiGHS (verified locally) in
PR 19255. However it takes very long to run and was already near the
limit for CircleCI. The tracking issue for this already existed
as well: scipy#19389

[skip actions] [skip cirrus]

mdhaber committed:
TST: optimize.linprog/milp: add tests for various bug reports

[skip ci]

HaoZeke committed:
TST: Mark xfail for upstream bugs

Co-authored-by: mdhaber <mdhaber@users.noreply.github.com>

HaoZeke committed:
ENH: Rework to drop cython highs_constants

HaoZeke committed:
MAINT: Remove older cython wrappers to HiGHs

HaoZeke committed:
ENH: Generate new highs_wrapper

Based on the internal Pybind11 bindings

HaoZeke committed:
BLD: Cleanup and prepare for new highs

TODO: Switch to scipy/highs

MAINT: Remove print debugging

HaoZeke committed:
BLD: Back to gitmodules for HiGHs

HaoZeke committed:
DOC: Add a note on submodule location

Co-authored-by: h-vetinari <h-vetinari@users.noreply.github.com>

HaoZeke committed:
TST: xfail known flaky tests

Co-authored-by: mckib2 <mckib2@users.noreply.github.com>

TST: Mark a 32-bit buggy test for HiGHS

HaoZeke committed:
BLD: Rework for CI errors

MAINT: Fixup mypy

BLD: Silence more warnings from highs

MAINT: Update highs commit

MAINT: Bump highs commit

BLD: Rework to have more control over highs build

MAINT: Try fixing mypy again

HaoZeke commented:
As may be seen in https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770 there's still the too many public symbols issue:

<details>
<summary> Failing build snippet </summary>

```
~/work/scipy/scipy/tools ~/work/scipy/scipy
----------- All the test files were installed --------------
----------- All the necessary .pyi files were installed --------------
../build/scipy/optimize/_highs/_core.cpython-310-x86_64-linux-gnu.so: too many public symbols!
00000000000598d0 T PyInit__core
000000000021b6c0 W _ZN9__gnu_cxx12__to_xstringINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEcEET_PFiPT0_mPKS8_P13__va_list_tagEmSB_z
0000000000419a70 W _ZNKSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_16LpSectionKeywordESaIS9_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSB_18_Mod_range_hashingENSB_20_Default_ranged_hashENSB_20_Prime_rehash_policyENSB_[17](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:18)_Hashtable_traitsILb1ELb0ELb1EEEE19_M_find_before_nodeEmRS7_m
000000000012f1b0 W _ZNKSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_[18](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:19)_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE[19](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:20)_M_find_before_nodeEmRS7_m
00000000003258b0 W _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7compareEPKc
0000000000419560 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_16LpSectionKeywordESaIS9_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSB_18_Mod_range_hashingENSB_[20](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:21)_Default_ranged_hashENSB_20_Prime_rehash_policyENSB_17_Hashtable_traitsILb1ELb0ELb1EEEE5clearEv
0000000000419b10 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_16LpSectionKeywordESaIS9_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSB_18_Mod_range_hashingENSB_20_Default_ranged_hashENSB_20_Prime_rehash_policyENSB_17_Hashtable_traitsILb1ELb0ELb1EEEEC1IPKS9_EET_SQ_mRKSG_RKSE_RKSA_St17integral_constantIbLb1EE
0000000000419b10 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_16LpSectionKeywordESaIS9_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSB_18_Mod_range_hashingENSB_20_Default_ranged_hashENSB_20_Prime_rehash_policyENSB_17_Hashtable_traitsILb1ELb0ELb1EEEEC2IPKS9_EET_SQ_mRKSG_RKSE_RKSA_St17integral_constantIbLb1EE
00000000000845d0 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_PvESaIS9_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSB_18_Mod_range_hashingENSB_20_Default_ranged_hashENSB_20_Prime_rehash_policyENSB_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm
00000000000a3510 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE18_M_assign_elementsIRKSL_EEvOT_
000000000012f3e0 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE[21](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:22)_M_insert_unique_nodeEmmPNSA_10_Hash_nodeIS8_Lb1EEEm
000000000010dc90 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE4findERS7_
0000000000066130 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE5clearEv
00000000000a31f0 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_assignIRKSL_NSA_17_ReuseOrAllocNodeISaINSA_10_Hash_nodeIS8_Lb1EEEEEEEEvOT_RKT0_
000000000012f2a0 W _ZNSt10_HashtableINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_iESaIS8_ENSt8__detail10_Select1stESt8equal_toIS5_ESt4hashIS5_ENSA_18_Mod_range_hashingENSA_20_Default_ranged_hashENSA_20_Prime_rehash_policyENSA_17_Hashtable_traitsILb1ELb0ELb1EEEE9_M_rehashEmRKm
0000000000393fa0 W _ZNSt10_HashtableImSt4pairIKmiESaIS2_ENSt8__detail10_Select1stESt8equal_toImESt4hashImENS4_18_Mod_range_hashingENS4_20_Default_ranged_hashENS4_20_Prime_rehash_policyENS4_17_Hashtable_traitsILb0ELb0ELb0EEEE11equal_rangeERS1_
0000000000219280 W _ZNSt10_HashtableImSt4pairIKmiESaIS2_ENSt8__detail10_Select1stESt8equal_toImESt4hashImENS4_18_Mod_range_hashingENS4_20_Default_ranged_hashENS4_20_Prime_rehash_policyENS4_17_Hashtable_traitsILb0ELb0ELb0EEEE4findERS1_
00000000002182e0 W _ZNSt10_HashtableImSt4pairIKmiESaIS2_ENSt8__detail10_Select1stESt8equal_toImESt4hashImENS4_18_Mod_range_hashingENS4_20_Default_ranged_hashENS4_20_Prime_rehash_policyENS4_17_Hashtable_traitsILb0ELb0ELb0EEEE5clearEv
000000000021b540 W _ZNSt10_HashtableImSt4pairIKmiESaIS2_ENSt8__detail10_Select1stESt8equal_toImESt4hashImENS4_18_Mod_range_hashingENS4_20_Default_ranged_hashENS4_20_Prime_rehash_policyENS4_17_Hashtable_traitsILb0ELb0ELb0EEEE5eraseENS4_20_Node_const_iteratorIS2_Lb0ELb0EEE
000000000024e250 W _ZNSt6vectorI11BasisStatusSaIS0_EE17_M_realloc_insertIJS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
0000000000140540 W _ZNSt6vectorI12HighsVarTypeSaIS0_EE14_M_fill_assignEmRKS0_
0000000000325920 W _ZNSt6vectorI12HighsVarTypeSaIS0_EE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPS0_S2_EEmRKS0_
000000000010d7a0 W _ZNSt6vectorI12HighsVarTypeSaIS0_EE17_M_default_appendEm
00000000000e83a0 W _ZNSt6vectorI12HighsVarTypeSaIS0_EE17_M_realloc_insertIJRKS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
0000000000341ab0 W _ZNSt6vectorI12HighsVarTypeSaIS0_EE17_M_realloc_insertIJS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
00000000001296c0 W _ZNSt6vectorI16HighsBasisStatusSaIS0_EE14_M_fill_assignEmRKS0_
0000000000129800 W _ZNSt6vectorI16HighsBasisStatusSaIS0_EE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPS0_S2_EEmRKS0_
000000000010dd50 W _ZNSt6vectorI16HighsBasisStatusSaIS0_EE17_M_default_appendEm
00000000000e9280 W _ZNSt6vectorI16HighsBasisStatusSaIS0_EE17_M_realloc_insertIJRKS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
000000000016b040 W _ZNSt6vectorI16HighsBasisStatusSaIS0_EE17_M_realloc_insertIJS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
00000000001d3d60 W _ZNSt6vectorIN18HighsTransformedLp9BoundTypeESaIS1_EE17_M_default_appendEm
0000000000341940 W _ZNSt6vectorIN18free_format_parser6HMpsFF8ConeTypeESaIS2_EE17_M_realloc_insertIJRKS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_
0000000000341c00 W _ZNSt6vectorIN18free_format_parser6HMpsFF9BoundtypeESaIS2_EE17_M_realloc_insertIJS2_EEEvN9__gnu_cxx17__normal_iteratorIPS2_S4_EEDpOT_
00000000004044a0 W _ZNSt6vectorIN3ipx7Iterate11StateDetailESaIS2_EE17_M_default_appendEm
000000000010d960 W _ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE17_M_default_appendEm
0000000000113d50 W _ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE17_M_realloc_insertIJRKS5_EEEvN9__gnu_cxx17__normal_iteratorIPS5_S7_EEDpOT_
00000000000831f0 W _ZNSt6vectorINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESaIS5_EE17_M_realloc_insertIJS5_EEEvN9__gnu_cxx17__normal_iteratorIPS5_S7_EEDpOT_
00000000003021f0 W _ZNSt6vectorIPSt15basic_streambufIcSt11char_traitsIcEESaIS4_EE17_M_realloc_insertIJRKS4_EEEvN9__gnu_cxx17__normal_iteratorIPS4_S6_EEDpOT_
0000000000065ce0 W _ZNSt6vectorIPcSaIS0_EE17_M_realloc_insertIJRKS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
0000000000239530 W _ZNSt6vectorIPiSaIS0_EE17_M_default_appendEm
000000000023ac00 W _ZNSt6vectorIPiSaIS0_EE17_M_realloc_insertIJS0_EEEvN9__gnu_cxx17__normal_iteratorIPS0_S2_EEDpOT_
00000000003423d0 W _ZNSt6vectorIS_ISt4pairIidESaIS1_EESaIS3_EE17_M_realloc_insertIJS3_EEEvN9__gnu_cxx17__normal_iteratorIPS3_S5_EEDpOT_
0000000000341760 W _ZNSt6vectorIS_ISt5tupleIJiidEESaIS1_EESaIS3_EE17_M_default_appendEm
000000000025fdc0 W _ZNSt6vectorIS_IdSaIdEESaIS1_EE14_M_fill_assignEmRKS1_
0000000000342090 W _ZNSt6vectorIS_IiSaIiEESaIS1_EE17_M_realloc_insertIJS1_EEEvN9__gnu_cxx17__normal_iteratorIPS1_S3_EEDpOT_
000000000018fa60 W _ZNSt6vectorISt23_Rb_tree_const_iteratorISt4pairIdlEESaIS3_EE17_M_default_appendEm
0000000000218e20 W _ZNSt6vectorISt23_Rb_tree_const_iteratorISt4pairIiiEESaIS3_EE14_M_fill_assignEmRKS3_
000000000021b890 W _ZNSt6vectorISt3setIiSt4lessIiESaIiEESaIS4_EE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPS4_S6_EEmRKS4_
00000000002188f0 W _ZNSt6vectorISt3setIiSt4lessIiESaIiEESaIS4_EE17_M_default_appendEm
00000000003de970 W _ZNSt6vectorISt4pairIS_IiSaIiEES_IdSaIdEEESaIS5_EE12emplace_backIJRS2_RS4_EEERS5_DpOT_
00000000003de580 W _ZNSt6vectorISt4pairIS_IiSaIiEES_IdSaIdEEESaIS5_EE17_M_realloc_insertIJRS2_RS4_EEEvN9__gnu_cxx17__normal_iteratorIPS5_S7_EEDpOT_
0000000000297ec0 W _ZNSt6vectorISt4pairIdiESaIS1_EE17_M_default_appendEm
0000000000129d40 W _ZNSt6vectorISt4pairIdiESaIS1_EE17_M_realloc_insertIJS1_EEEvN9__gnu_cxx17__normal_iteratorIPS1_S3_EEDpOT_
00000000002a9390 W _ZNSt6vectorISt4pairIdiESaIS1_EE7reserveEm
00000000003de290 W _ZNSt6vectorISt4pairIidESaIS1_EE12emplace_backIJRKiRKdEEERS1_DpOT_
00000000003d6bb0 W _ZNSt6vectorISt4pairIidESaIS1_EE12emplace_backIJRiRdEEERS1_DpOT_
000000000029c400 W _ZNSt6vectorISt4pairIidESaIS1_EE17_M_default_appendEm
0000000000341d70 W _ZNSt6vectorISt4pairIidESaIS1_EE17_M_realloc_insertIJS1_EEEvN9__gnu_cxx17__normal_iteratorIPS1_S3_EEDpOT_
00000000003de1c0 W _ZNSt6vectorISt4pairIidESaIS1_EE7reserveEm
00000000003af350 W _ZNSt6vectorISt4pairIiiESaIS1_EE14_M_fill_assignEmRKS1_
00000000003aec80 W _ZNSt6vectorISt4pairIiiESaIS1_EE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPS1_S3_EEmRKS1_
00000000002d6620 W _ZNSt6vectorISt4pairIiiESaIS1_EE17_M_realloc_insertIJS1_EEEvN9__gnu_cxx17__normal_iteratorIPS1_S3_EEDpOT_
0000000000218dc0 W _ZNSt6vectorISt4pairIiiESaIS1_EE8_M_eraseEN9__gnu_cxx17__normal_iteratorIPS1_S3_EES7_
0000000000239870 W _ZNSt6vectorISt4pairIijESaIS1_EE17_M_default_appendEm
0000000000341ee0 W _ZNSt6vectorISt5tupleIJiidEESaIS1_EE17_M_realloc_insertIJS1_EEEvN9__gnu_cxx17__normal_iteratorIPS1_S3_EEDpOT_
000000000010f080 W _ZNSt6vectorISt8functionIFvR13QpModelStatusEESaIS4_EE17_M_realloc_insertIJRKS4_EEEvN9__gnu_cxx17__normal_iteratorIPS4_S6_EEDpOT_
000000000010f580 W _ZNSt6vectorISt8functionIFvRiEESaIS3_EE17_M_realloc_insertIJRKS3_EEEvN9__gnu_cxx17__normal_iteratorIPS3_S5_EEDpOT_
00000000001b0380 W _ZNSt6vectorISt8multimapIdiSt4lessIdESaISt4pairIKdiEEESaIS7_EE17_M_default_appendEm
0000000000129a90 W _ZNSt6vectorIaSaIaEE14_M_fill_assignEmRKa
00000000001294f0 W _ZNSt6vectorIaSaIaEE17_M_default_appendEm
0000000000219130 W _ZNSt6vectorIaSaIaEE17_M_realloc_insertIJRKaEEEvN9__gnu_cxx17__normal_iteratorIPaS1_EEDpOT_
00000000003731f0 W _ZNSt6vectorIaSaIaEE17_M_realloc_insertIJaEEEvN9__gnu_cxx17__normal_iteratorIPaS1_EEDpOT_
000000000007ad50 W _ZNSt6vectorIbSaIbEE14_M_fill_insertESt13_Bit_iteratormb
00000000002d64e0 W _ZNSt6vectorIcSaIcEE14_M_fill_assignEmRKc
0000000000219420 W _ZNSt6vectorIcSaIcEE17_M_default_appendEm
0000000000334db0 W _ZNSt6vectorIcSaIcEE17_M_realloc_insertIJRKcEEEvN9__gnu_cxx17__normal_iteratorIPcS1_EEDpOT_
0000000000349540 W _ZNSt6vectorIcSaIcEE17_M_realloc_insertIJcEEEvN9__gnu_cxx17__normal_iteratorIPcS1_EEDpOT_
000000000032c560 W _ZNSt6vectorIdSaIdEE12emplace_backIJdEEERdDpOT_
00000000003d7190 W _ZNSt6vectorIdSaIdEE13_M_assign_auxIPKdEEvT_S5_St20forward_iterator_tag
000000000010db40 W _ZNSt6vectorIdSaIdEE14_M_fill_assignEmRKd
000000000010d360 W _ZNSt6vectorIdSaIdEE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPdS1_EEmRKd
000000000010c2e0 W _ZNSt6vectorIdSaIdEE17_M_default_appendEm
0000000000113be0 W _ZNSt6vectorIdSaIdEE17_M_realloc_insertIJRKdEEEvN9__gnu_cxx17__normal_iteratorIPdS1_EEDpOT_
0000000000083080 W _ZNSt6vectorIdSaIdEE17_M_realloc_insertIJdEEEvN9__gnu_cxx17__normal_iteratorIPdS1_EEDpOT_
0000000000217650 W _ZNSt6vectorIdSaIdEE7reserveEm
00000000002187b0 W _ZNSt6vectorIhSaIhEE14_M_fill_assignEmRKh
0000000000170c60 W _ZNSt6vectorIhSaIhEE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPhS1_EEmRKh
0000000000171070 W _ZNSt6vectorIhSaIhEE17_M_default_appendEm
00000000003b2f60 W _ZNSt6vectorIhSaIhEE17_M_realloc_insertIJhEEEvN9__gnu_cxx17__normal_iteratorIPhS1_EEDpOT_
000000000032c7b0 W _ZNSt6vectorIiSaIiEE12emplace_backIJiEEERiDpOT_
00000000003d7060 W _ZNSt6vectorIiSaIiEE13_M_assign_auxIPKiEEvT_S5_St20forward_iterator_tag
000000000010d660 W _ZNSt6vectorIiSaIiEE14_M_fill_assignEmRKi
0000000000140680 W _ZNSt6vectorIiSaIiEE14_M_fill_insertEN9__gnu_cxx17__normal_iteratorIPiS1_EEmRKi
000000000010c110 W _ZNSt6vectorIiSaIiEE17_M_default_appendEm
0000000000113a70 W _ZNSt6vectorIiSaIiEE17_M_realloc_insertIJRKiEEEvN9__gnu_cxx17__normal_iteratorIPiS1_EEDpOT_
0000000000082f10 W _ZNSt6vectorIiSaIiEE17_M_realloc_insertIJiEEEvN9__gnu_cxx17__normal_iteratorIPiS1_EEDpOT_
0000000000217010 W _ZNSt6vectorIiSaIiEE7reserveEm
0000000000187870 W _ZNSt6vectorIiSaIiEE8_M_eraseEN9__gnu_cxx17__normal_iteratorIPiS1_EES5_
00000000003f2ee0 W _ZNSt6vectorIiSaIiEEaSERKS1_
00000000003d8820 W _ZNSt6vectorIjSaIjEE14_M_fill_assignEmRKj
000000000023a6e0 W _ZNSt6vectorIjSaIjEE17_M_default_appendEm
00000000002393c0 W _ZNSt6vectorIjSaIjEE17_M_realloc_insertIJRKjEEEvN9__gnu_cxx17__normal_iteratorIPjS1_EEDpOT_
000000000032c810 W _ZNSt6vectorIjSaIjEE17_M_realloc_insertIJjEEEvN9__gnu_cxx17__normal_iteratorIPjS1_EEDpOT_
0000000000238580 W _ZNSt6vectorIjSaIjEE7reserveEm
00000000003d7430 W _ZNSt6vectorIlSaIlEE12emplace_backIJlEEERlDpOT_
00000000001904b0 W _ZNSt6vectorIlSaIlEE17_M_realloc_insertIJRKlEEEvN9__gnu_cxx17__normal_iteratorIPlS1_EEDpOT_
00000000003d72c0 W _ZNSt6vectorIlSaIlEE17_M_realloc_insertIJlEEEvN9__gnu_cxx17__normal_iteratorIPlS1_EEDpOT_
000000000037a640 W _ZNSt6vectorIsSaIsEE17_M_default_appendEm
000000000034[22](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:23)60 W _ZNSt6vectorIsSaIsEE17_M_realloc_insertIJsEEEvN9__gnu_cxx17__normal_iteratorIPsS1_EEDpOT_
0000000000218670 W _ZNSt6vectorItSaItEE14_M_fill_assignEmRKt
0000000000218f60 W _ZNSt6vectorItSaItEE17_M_default_appendEm
000000000006d9f0 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm
0000000000064640 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructEmc
000000000021b600 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12_M_constructIPcEEvT_S7_St20forward_iterator_tag
0000000000418c90 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7reserveEv
000000000006a9a0 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_appendEPKcm
0000000000069900 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_assignERKS4_
000000000006a830 W _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_mutateEmmPKcm
000000000032c980 W _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_jESt10_Select1stIS8_ESt4lessIS5_ESaIS8_EE24_M_get_insert_unique_posERS7_
000000000032caf0 W _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_jESt10_Select1stIS8_ESt4lessIS5_ESaIS8_EE29_M_get_insert_hint_unique_posESt[23](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:24)_Rb_tree_const_iteratorIS8_ERS7_
000000000032bdf0 W _ZNSt8_Rb_treeINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS5_jESt10_Select1stIS8_ESt4lessIS5_ESaIS8_EE8_M_eraseEPSt13_Rb_tree_nodeIS8_E
0000000000394040 W _ZNSt8_Rb_treeISt4pairIiiES1_St9_IdentityIS1_ESt4lessIS1_ESaIS1_EE11equal_rangeERKS1_
0000000000394120 W _ZNSt8_Rb_treeISt4pairIiiES1_St9_IdentityIS1_ESt4lessIS1_ESaIS1_EE12_M_erase_auxESt23_Rb_tree_const_iteratorIS1_ES9_
000000000037a4e0 W _ZNSt8_Rb_treeISt4pairIiiES1_St9_IdentityIS1_ESt4lessIS1_ESaIS1_EE17_M_emplace_uniqueIJiRiEEES0_ISt17_Rb_tree_iteratorIS1_EbEDpOT_
000000000036ce60 W _ZNSt8_Rb_treeISt4pairIiiES1_St9_IdentityIS1_ESt4lessIS1_ESaIS1_EE[24](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:25)_M_get_insert_unique_posERKS1_
00000000001b0710 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE12_M_erase_auxESt23_Rb_tree_const_iteratorIS2_ESA_
00000000001b0650 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE16_M_emplace_equalIJdRiEEESt17_Rb_tree_iteratorIS2_EDpOT_
00000000002192f0 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE17_M_emplace_uniqueIJRdiEEES0_ISt17_Rb_tree_iteratorIS2_EbEDpOT_
00000000001b0950 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE21_M_emplace_hint_equalIJRdRiEEESt17_Rb_tree_iteratorIS2_ESt23_Rb_tree_const_iteratorIS2_EDpOT_
0000000000334f00 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE24_M_get_insert_unique_posERS1_
00000000001b07b0 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE28_M_get_insert_hint_equal_posESt23_Rb_tree_const_iteratorIS2_ERS1_
0000000000334fb0 W _ZNSt8_Rb_treeIdSt4pairIKdiESt10_Select1stIS2_ESt4lessIdESaIS2_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorIS2_ERS1_
000000000023af50 W _ZNSt8_Rb_treeIdSt4pairIKdjESt10_Select1stIS2_ESt4lessIdESaIS2_EE22_M_emplace_hint_uniqueIJRdmEEESt17_Rb_tree_iteratorIS2_ESt23_Rb_tree_const_iteratorIS2_EDpOT_
000000000023ad70 W _ZNSt8_Rb_treeIdSt4pairIKdjESt10_Select1stIS2_ESt4lessIdESaIS2_EE24_M_get_insert_unique_posERS1_
000000000023ae20 W _ZNSt8_Rb_treeIdSt4pairIKdjESt10_Select1stIS2_ESt4lessIdESaIS2_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorIS2_ERS1_
0000000000[25](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:26)6310 W _ZNSt8_Rb_treeIiSt4pairIKi11BasisStatusESt10_Select1stIS3_ESt4lessIiESaIS3_EE24_M_get_insert_unique_posERS1_
00000000002563b0 W _ZNSt8_Rb_treeIiSt4pairIKi11BasisStatusESt10_Select1stIS3_ESt4lessIiESaIS3_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorIS3_ERS1_
00000000003f8560 W _ZNSt8_Rb_treeIiSt4pairIKiNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEESt10_Select1stIS8_ESt4lessIiESaIS8_EE24_M_get_insert_unique_posERS1_
00000000003f8600 W _ZNSt8_Rb_treeIiSt4pairIKiNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEESt10_Select1stIS8_ESt4lessIiESaIS8_EE29_M_get_insert_hint_unique_posESt23_Rb_tree_const_iteratorIS8_ERS1_
0000000000238eb0 W _ZNSt8_Rb_treeIiiSt9_IdentityIiESt4lessIiESaIiEE16_M_insert_uniqueIRKiEESt4pairISt17_Rb_tree_iteratorIiEbEOT_
0000000000218bc0 W _ZNSt8_Rb_treeIiiSt9_IdentityIiESt4lessIiESaIiEE17_M_emplace_uniqueIJRiEEESt4pairISt17_Rb_tree_iteratorIiEbEDpOT_
000000000018fe00 W _ZNSt8_Rb_treeIllSt9_IdentityIlESt4lessIlESaIlEE16_M_insert_uniqueIRKlEESt4pairISt17_Rb_tree_iteratorIlEbEOT_
0000000000062d30 W _ZNSt8__detail18__to_chars_10_implIjEEvPcjT_
0000000000084710 W _ZNSt8__detail9_Map_baseINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt4pairIKS6_PvESaISA_ENS_10_Select1stESt8equal_toIS6_ESt4hashIS6_ENS_18_Mod_range_hashingENS_20_Default_ranged_hashENS_20_Prime_rehash_policyENS_17_Hashtable_traitsILb1ELb0ELb1EEELb1EEixEOS6_
0000000000218cc0 W _ZSt11__remove_ifIN9__gnu_cxx17__normal_iteratorIPiSt6vectorIiSaIiEEEENS0_5__ops16_Iter_equals_valIKiEEET_SB_SB_T0_
0000000000420690 W _ZSt15__valarray_copyIdNSt8__detail8_BinClosISt6__plusSt9_ValArraySt5_ExprdNS1_ISt12__multipliesSt9_ConstantS3_ddEEEEEvRKS4_IT0_T_EmSt6_ArrayISA_E
00000000004205b0 W _ZSt23_Array_augmented___plusIdNSt8__detail8_BinClosISt12__multipliesSt9_ConstantSt9_ValArrayddEEEvSt6_ArrayIT_ERKSt5_ExprIT0_S7_Em
0000000000420620 W _ZSt24_Array_augmented___minusIdNSt8__detail8_BinClosISt12__multipliesSt9_ConstantSt9_ValArrayddEEEvSt6_ArrayIT_ERKSt5_ExprIT0_S7_Em
000000000006aa90 W _ZStplIcSt11char_traitsIcESaIcEENSt7__cxx1112basic_stringIT_T0_T1_EEPKS5_RKS8_
000000000050dd40 V _ZTI11HighsStatus
000000000050dd60 V _ZTI11IisStrategy
000000000050dd50 V _ZTI12HighsLogType
000000000050dd10 V _ZTI12HighsVarType
000000000050dcc0 V _ZTI12MatrixFormat
000000000050dc90 V _ZTI13BasisValidity
000000000050dcd0 V _ZTI13HessianFormat
000000000050dd30 V _ZTI13HighsInfoType
000000000050de10 V _ZTI14EdgeWeightMode
000000000050dd70 V _ZTI14IisBoundStatus
000000000050dc80 V _ZTI14SolutionStatus
000000000050dca0 V _ZTI15HighsDebugLevel
000000000050dd20 V _ZTI15HighsOptionType
000000000050dd80 V _ZTI15SimplexStrategy
000000000050dd00 V _ZTI16HighsBasisStatus
000000000050dce0 V _ZTI16HighsModelStatus
000000000050de20 V _ZTI17HighsCallbackType
000000000050ddb0 V _ZTI17SimplexSolvePhase
000000000050dcf0 V _ZTI19HighsPresolveStatus
000000000050de00 V _ZTI19SimplexNlaOperation
000000000050dd90 V _ZTI20SimplexCrashStrategy
000000000050ddd0 V _ZTI20SimplexPriceStrategy
000000000050ddc0 V _ZTI25SimplexEdgeWeightStrategy
000000000050ddf0 V _ZTI31SimplexPrimalCorrectionStrategy
000000000050dda0 V _ZTI31SimplexUnscaledSolutionStrategy
000000000050dde0 V _ZTI35SimplexPivotalRowRefinementStrategy
000000000050dcb0 V _ZTI8ObjSense
000000000050f320 V _ZTINSt6thread11_State_implINS_8_InvokerISt5tupleIJZN17HighsTaskExecutorC4EiEUliE_iEEEEEE
000000000050d6c8 V _ZTISt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
000000000050d988 V _ZTISt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE
000000000043a8d0 V _ZTS11HighsStatus
000000000043a8f0 V _ZTS11IisStrategy
000000000043a8e0 V _ZTS12HighsLogType
000000000043a888 V _ZTS12HighsVarType
000000000043a808 V _ZTS12MatrixFormat
000000000043a7d0 V _ZTS13BasisValidity
000000000043a820 V _ZTS13HessianFormat
000000000043a8c0 V _ZTS13HighsInfoType
000000000043aa90 V _ZTS14EdgeWeightMode
000000000043a900 V _ZTS14IisBoundStatus
000000000043a7b0 V _ZTS14SolutionStatus
000000000043a7e0 V _ZTS15HighsDebugLevel
000000000043a8a0 V _ZTS15HighsOptionType
000000000043a920 V _ZTS15SimplexStrategy
000000000043a870 V _ZTS16HighsBasisStatus
000000000043a830 V _ZTS16HighsModelStatus
000000000043aab0 V _ZTS17HighsCallbackType
000000000043a990 V _ZTS17SimplexSolvePhase
000000000043a850 V _ZTS19HighsPresolveStatus
000000000043aa70 V _ZTS19SimplexNlaOperation
000000000043a940 V _ZTS20SimplexCrashStrategy
000000000043a9d0 V _ZTS20SimplexPriceStrategy
000000000043a9b0 V _ZTS25SimplexEdgeWeightStrategy
000000000043aa40 V _ZTS31SimplexPrimalCorrectionStrategy
000000000043a960 V _ZTS31SimplexUnscaledSolutionStrategy
000000000043aa00 V _ZTS35SimplexPivotalRowRefinementStrategy
000000000043a7f8 V _ZTS8ObjSense
000000000043b92a V _ZTSA1_c
0000000000444540 V _ZTSNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN17HighsTaskExecutorC4EiEUliE_iEEEEEE
000000000043b800 V _ZTSNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE
0000000000439600 V _ZTSSt11_Mutex_baseILN9__gnu_cxx12_Lock_policyE2EE
000000000043a400 V _ZTSSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE
00000000004501c0 V _ZTSSt19_Sp_make_shared_tag
000000000043b930 V _ZTSSt6vectorIdSaIdEE
000000000050f3a8 V _ZTVNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN17HighsTaskExecutorC4EiEUliE_iEEEEEE
0000000000450290 u _ZZNSt19_Sp_make_shared_tag5_S_tiEvE5__tag
000000000043e980 u _ZZNSt8__detail18__to_chars_10_implIjEEvPcjT_E8__digits
000000000043e8a0 u _ZZNSt8__detail18__to_chars_10_implImEEvPcjT_E8__digits
0000000000077d00 W _ZZNSt9once_flag18_Prepare_executionC4IZSt9call_onceIZN8pybind1128gil_safe_call_once_and_storeINS3_6detail7npy_apiEE[26](https://github.com/HaoZeke/scipy/actions/runs/10879376567/job/30183963770#step:11:27)call_once_and_store_resultIRFS6_vEEERS7_OT_EUlvE_JEEvRS_SD_DpOT0_EUlvE_EERSC_ENUlvE_4_FUNEv
Error: Process completed with exit code 1.
```
</details>

Which is superficially similar to https://github.com/scipy/scipy/pull/19255#issuecomment-1742169471 but I can't seem to figure out a fix.

lucascolley commented on scipy/optimize/_highs/meson.build:
```diff
@@ -1,290 +1,75 @@
-  '-UOSI_FOUND',
-  '-DNDEBUG'
-]
+fs = import('fs')
+if not fs.exists('../../../subprojects/highs/README.md')
+    error(
+        'Missing the `highs` submodule! Run `git submodule update --init` to fix this.',
+    )
+endif
```
is this needed? I think Meson should error if this is the case and highs is not a system library (https://mesonbuild.com/Reference-manual_functions.html#subproject)

HaoZeke replied:
Yes, but I believe this was done to keep the build system consistent, i.e. even people unfamiliar with `meson` should be able to continue to get a working checkout with the old submodule instructions.

I had initially added a `.wrap` file since that's the more `meson`-centric approach (and easier to handle IMO) but there was a request to bring back the submodules (and more context as to why) over in https://github.com/scipy/scipy/pull/19255#issuecomment-1745605754 and subsequent messages.

lucascolley replied:
thanks, I suppose that answers my question at https://github.com/mesonbuild/meson/discussions/13683. Looks like it boils down to "it doesn't really matter which you use, but staying with a submodule is probably less disruption for SciPy".

In gh-21270, I need a wrap file anyway to invoke CMake, so I wonder whether having a submodule as well is still preferable to using `[wrap-git]`.

lucascolley replied:
we'd have to remove this if we wanted to support using highs as a system library though, right?

HaoZeke replied:
Yup, or rather make it conditional on another meson option. I don't think we're likely to allow that anytime soon though, highs moves far too quickly, which is why we vendor it under Scipy and also why we don't directly import highspy

ev-br commented:
> here's still the too many public symbols issue:

Try adding `link args: version_link_args` to meson build. Seemed to do the trick in a recent PR:

https://github.com/scipy/scipy/pull/21551/commits/63569e0d2acb751ef396137a29ba67400ef8b1f0

ev-br commented:
And indeed, applying

```diff
$ git diff
diff --git a/scipy/optimize/_highs/meson.build b/scipy/optimize/_highs/meson.build
index cc29ea2bf2..38fde6217c 100644
--- a/scipy/optimize/_highs/meson.build
+++ b/scipy/optimize/_highs/meson.build
@@ -58,6 +58,7 @@ py3.extension_module(
     dependencies: scipy_highspy_dep,
     c_args: [Wno_unused_variable, Wno_unused_but_set_variable],
     cpp_args: [_cpp_Wno_unused_variable, _cpp_Wno_unused_but_set_variable],
+    link_args: version_link_args,
     gnu_symbol_visibility: highs_symbol_visibility,
     install: true,
     subdir: 'scipy/optimize/_highspy',
@@ -69,6 +70,7 @@ py3.extension_module(
     dependencies: scipy_highspy_dep,
     c_args: [Wno_unused_variable, Wno_unused_but_set_variable],
     cpp_args: [_cpp_Wno_unused_variable, _cpp_Wno_unused_but_set_variable],
+    link_args: version_link_args,
     gnu_symbol_visibility: highs_symbol_visibility,
     install: true,
     subdir: 'scipy/optimize/_highspy',
```

seems to fix the symbol-hiding check locally:

```
$ ./tools/check_pyext_symbol_hiding.sh build/scipy/optimize/_highs/
Symbol hiding OK
```

HaoZeke committed:
BLD: Fix symbol visibility issues

Co-authored-by: ev-br <ev-br@users.noreply.github.com>

ev-br commented:
The smoke-tutorials CI failure looks related.

HaoZeke committed:
DOC: Fix incorrect primal_status

The primal status should not be at a lower / fixed bound, that should be
from the basis status for each row and column

aa = highs.getBasis()
[highs.basisStatusToString(x) fox x in aa.row_status]
['At upper bound', 'At upper bound', 'At lower/fixed bound', 'At
lower/fixed bound']
[highs.basisStatusToString(x) fox x in aa.col_status]
['Basic', 'Basic', 'Basic, 'At lower/fixed bound']

HaoZeke commented:
> The smoke-tutorials CI failure looks related.

Yup, weirdly long-standing bug in the documentation, the `primal_status` is indeed `None` there :)

HaoZeke committed:
MAINT: Update highs again

HaoZeke committed:
MAINT: Minor formatting

HaoZeke committed:
MAINT: Fix more warnings

HaoZeke commented:
Upstream has expressed an interest in not directly having to maintain `meson.build` in their repo as they have chosen to go with `cmake` / `scikit-build`, see https://github.com/ERGO-Code/HiGHS/pull/1987#pullrequestreview-2380668057.

Given this, we could either:
- Vendor the `meson.build` files only in the SciPy fork
- Let the `meson.build` files be removed from both upstream and the fork, but keep them in `wrapdb`

I would prefer the second option since it is a bit cleaner in my opinion, however, AFAIK `wrapdb` doesn't support versioning of `meson.build` files, so maybe keeping the build in the fork is best..

HaoZeke commented:
Also, accelerate CI failure seems unrelated:

```
scipy/meson.build:244:11: ERROR: Dependency "OpenBLAS" not found, tried pkgconfig, framework and cmake
```

`asv` benchmark [failure on Circle-CI](https://app.circleci.com/pipelines/github/scipy/scipy/30744/workflows/4038bd0f-7d9e-4623-b1ae-65ba65c2cb87/jobs/103931/parallel-runs/0/steps/0-104) seems real, investigating.

```bash
[22.36%] ··· optimize_linprog.Netlib.time_netlib                   1/184 failed
[22.36%] ··· =================== ========== ==========
                    method        problems            
             ------------------- ---------- ----------
...
              ('highs-ipm', {})    GROW7      failed  
             For parameters: ('highs-ipm', {}), 'GROW7'
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:162: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.c = data["c"]
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:163: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.A_eq = data["A_eq"]
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:164: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.A_ub = data["A_ub"]
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:165: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.b_ub = data["b_ub"]
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:166: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.b_eq = data["b_eq"]
             /home/circleci/repo/benchmarks/benchmarks/optimize_linprog.py:167: UserWarning: Reading `.npy` or `.npz` file required additional header parsing as it was created on Python 2. Save the file again to speed up loading and avoid this warning.
               self.bounds = np.squeeze(data["bounds"])
             python3.11: ../subprojects/highs/src/lp_data/HighsSolution.cpp:291: void getKktFailures(const HighsOptions&, const HighsLp&, const std::vector<double>&, const HighsSolution&, const HighsBasis&, HighsInfo&, HighsPrimalDualErrors&, bool): Assertion `status_value_ok' failed.
```

rgommers commented:
> Upstream has expressed an interest in not directly having to maintain `meson.build` in their repo

It's not a done deal. I'd strongly prefer keeping them in-tree. It's mainly a matter of PR hygiene as far as I can tell - if we make it as easy as possible for them to keep it around and we make it a win-win by sending them useful fixes to their C++ code and Python bindings, then it should be fine I'd expect.

WrapDB-only is not an option, since we cannot rely on a download at build time. So the fallback option if HiGHS does remove `meson.build` files would be in our fork.

HaoZeke committed:
BLD: Rework to incorporate local highs_options

HaoZeke committed:
ENH: Add the mapping code into scipy

Since it is only used by these bindings

HaoZeke committed:
MAINT: Rework to highspy and highs API names

HaoZeke committed:
TMP: Switch to using the HiGHs branch

Later this commit shall be removed in favor of a branch of the fork rebased onto this...

HaoZeke committed:
MAINT: Pass linter with highspy C++ API

HaoZeke commented:
The `asv` issue seems to have resolved itself locally at any rate.

```bash
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[ 0.00%] ·· Benchmarking existing-py_home_rgoswami_micromamba_envs_scipy-dev_bin_python
[50.00%] ··· ...mize_linprog.Netlib.time_netlib                 ok
```

I'll wait for upstream https://github.com/ERGO-Code/HiGHS/pull/2005 before working on the warnings.

HaoZeke committed:
MAINT: Remove unused simplex_crash_strategy

As noted by @jajhall, these no longer do anything within HiGHS [1]

[1] https://github.com/ERGO-Code/HiGHS/pull/2005#pullrequestreview-2398423172

rgommers commented on scipy/optimize/_highs/meson.build:
```diff
@@ -1,290 +1,75 @@
+
+py3.install_sources(
+    ['__init__.py', '_highs_wrapper.py'],
+    subdir: 'scipy/optimize/_highspy',
```
Why have the directory be `optimize/_highs/` and install everything into `optimize/_highspy/`? That seems unnecessarily confusing, I'd choose one or the other.

HaoZeke replied:
Decided to go with `highspy` throughout.

rgommers commented:
This seems to be pretty close, just some last tweaks and build warnings to silence?

HaoZeke committed:
BLD: Use suppressions from HiGHS instead

HaoZeke committed:
BLD: Restructure to highspy consistently

Co-authored-by: rgommers <rgommers@users.noreply.github.com>

HaoZeke committed:
TMP: Update subproject commit

HaoZeke committed:
BLD: Use all highs arguments

HaoZeke commented:
The UCRT64 failure shouldn't be related to anything here (since none of the source files have changed, and the meson build isn't used in that CI job..)

HaoZeke commented:
Not sure what the problem with the CI benchmarks are, locally these still pass. Looking into it now.

HaoZeke committed:
BENCH: Disable GROW7 for now

HaoZeke commented:
The `mip` failure on 32-bit has cropped up a few times before:
- https://github.com/ERGO-Code/HiGHS/issues/769
- https://github.com/scipy/scipy/issues/16347

Most recently it seems to have been `xfail`'ed here:
- https://github.com/scipy/scipy/pull/16814

In general 32-bit support for highs is a very low priority:
- https://github.com/msys2/MINGW-packages/pull/13686
- https://github.com/ERGO-Code/HiGHS/issues/1110#issuecomment-1404171817

Which is OK, worst case might need to `git bisect` until a green commit is found.

HaoZeke committed:
TMP: Switch back to the fork with upstream latest

HaoZeke committed:
BLD: Try to use safer 32-bit FP for HiGHS

HaoZeke commented:
OK this is now good to go. The fix for the FP errors on 32-bit was to use `-ffloat-store` which is discussed here https://gcc.gnu.org/bugzilla/show_bug.cgi?id=323#c87

There is apparently a small performance penalty (only on 32-bit machines), but it makes it a lot more robust so this seems like the right option.

rgommers commented:
> Retarget to the SciPy fork (after the companion PR is merged)

When you do this, could you please add trigger wheel builds on this PR by including `[wheel build]` in the first line of the most recent commit you push?

rgommers commented on scipy/meson.build:
```diff
@@ -590,6 +591,7 @@ subdir('sparse')
 subdir('stats')
 subdir('fft')
 subdir('io')
+subdir('optimize')
```
Looking at build profiling for this change, looks good:

![image](https://github.com/user-attachments/assets/1a469701-77ab-424d-beac-e5b6351f1519)

HiGHS did get heavier I think, but that is what it is.

rgommers commented on scipy/optimize/_highspy/meson.build:
```diff
@@ -0,0 +1,82 @@
+    'do_install': false,
+}
+
+# Hacky workaround to not check cpu_family for 32 bit identifiers
```
This does look hacky indeed. Don't you want `cc.sizeof('size_t') == 4` here?

HaoZeke replied:
Decided to update with `sizeof(void*)` because of the minimum width of `size_t` being `16` thing [1]; also consistency, `PROPACK` uses the same check on `void*`.

[1] https://en.cppreference.com/w/c/types/size_t

rgommers commented on scipy/optimize/_highspy/meson.build:
```diff
@@ -0,0 +1,82 @@
+# Hacky workaround to not check cpu_family for 32 bit identifiers
+is_32bit = target_machine.cpu_family() in ['arc', 'arm', 'csky', 'ft32', 'mips', 'ppc', 'riscv32', 'rx', 'sparc', 'wasm32', 'x86']
+_float_store = cc.get_supported_arguments('-ffloat-store')
+_cpp_float_store = cpp.get_supported_arguments('-ffloat-store')
```
These can be moved inside the `if is_32bit`

rgommers commented on scipy/optimize/_highspy/meson.build:
```diff
@@ -0,0 +1,82 @@
+    include_directories: highs_incdirs,
+)
+
+scipy_highspy_dep = [py3_dep, pybind11_dep, _highs_dep, thread_dep, atomic_dep]
```
minor: no need to add `py3_dep`

rgommers commented on scipy/optimize/_highspy/meson.build:
```diff
@@ -0,0 +1,82 @@
+    sources: [highspy_cpp, highs_conf_file],
+    dependencies: scipy_highspy_dep,
+    link_args: version_link_args,
+    gnu_symbol_visibility: 'hidden',  # default
```
please remove - arguments that are only set to their default value aren't useful

rgommers commented on scipy/optimize/_highspy/meson.build:
```diff
@@ -0,0 +1,82 @@
+    sources: [highs_conf_file, 'highs_options.cpp'],
+    dependencies: scipy_highspy_dep,
+    link_args: version_link_args,
+    gnu_symbol_visibility: 'hidden',  # default
```
please remove here as well

rgommers reviewed (commented):
Looks quite good, only a few minor comments!

HaoZeke committed:
MAINT: Apply review comments for HiGHS interface

Co-authored-by: rgommers <rgommers@users.noreply.github.com>

HaoZeke committed:
MAINT: Use scipy fork of HiGHS [wheel build]

Now upstream compatible

HaoZeke commented:
I don't think the `threading` failure in [Oldest GCC & pydata/sparse, fast, py3.10/npMin, pip+pytest](https://github.com/scipy/scipy/actions/runs/11849072764/job/33021718223?pr=21565#logs) is related to changes here though I'll double check.

```bash
=================================== FAILURES ===================================
______________________ interpolate/tests/test_interpnd.py ______________________
[gw0] linux -- Python 3.10.15 /opt/hostedtoolcache/Python/3.10.15/x64/bin/python
worker 'gw0' crashed while running 'interpolate/tests/test_interpnd.py::TestLinearNDInterpolation::test_threading'
=============================== warnings summary ===============================
../../../../opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/_pytest/config/__init__.py:1278
../../../../opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/_pytest/config/__init__.py:1278
../../../../opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/_pytest/config/__init__.py:1278
../../../../opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/_pytest/config/__init__.py:1278
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/_pytest/config/__init__.py:1278: PytestConfigWarning: assertions not in test modules or plugins will be ignored because assert statements are not executed by the underlying Python interpreter (are you using python -O?)
  
    self._warn_about_missing_assertion(mode)

spatial/tests/test_distance.py: 15 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:3100: DeprecationWarning: The kulczynski1 metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'kulczynski1(u, v)' with '1/jaccard(u, v) - 1'.
    return cdist_fn(XA, XB, out=out, **kwargs)

spatial/tests/test_distance.py: 800 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2755: DeprecationWarning: The kulczynski1 metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'kulczynski1(u, v)' with '1/jaccard(u, v) - 1'.
    dm[i, j] = metric(XA[i], XB[j], **kwargs)

spatial/tests/test_distance.py: 15 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2295: DeprecationWarning: The kulczynski1 metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'kulczynski1(u, v)' with '1/jaccard(u, v) - 1'.
    return pdist_fn(X, out=out, **kwargs)

spatial/tests/test_distance.py: 380 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2744: DeprecationWarning: The kulczynski1 metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'kulczynski1(u, v)' with '1/jaccard(u, v) - 1'.
    dm[k] = metric(X[i], X[j], **kwargs)

spatial/tests/test_distance.py: 15 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:3100: DeprecationWarning: The sokalmichener metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'sokalmichener(u, v)' with 'rogerstanimoto(u, v)'.
    return cdist_fn(XA, XB, out=out, **kwargs)

spatial/tests/test_distance.py: 800 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2755: DeprecationWarning: The sokalmichener metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'sokalmichener(u, v)' with 'rogerstanimoto(u, v)'.
    dm[i, j] = metric(XA[i], XB[j], **kwargs)

spatial/tests/test_distance.py: 15 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2295: DeprecationWarning: The sokalmichener metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'sokalmichener(u, v)' with 'rogerstanimoto(u, v)'.
    return pdist_fn(X, out=out, **kwargs)

spatial/tests/test_distance.py: 380 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/spatial/distance.py:2744: DeprecationWarning: The sokalmichener metric is deprecated since SciPy 1.15.0 and will be removed in SciPy 1.17.0.  Replace usage of 'sokalmichener(u, v)' with 'rogerstanimoto(u, v)'.
    dm[k] = metric(X[i], X[j], **kwargs)

stats/tests/test_axis_nan_policy.py: 73 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/stats/_stats_py.py:10656: SmallSampleWarning: One or more sample arguments is too small; all returned values will be NaN. See documentation for sample size requirements.
    return gmean(x, weights=weights, axis=axis, keepdims=keepdims)

stats/tests/test_axis_nan_policy.py: 93 warnings
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/numpy/core/fromnumeric.py:3432: RuntimeWarning: Mean of empty slice.
    return _methods._mean(a, axis=axis, dtype=dtype,

stats/tests/test_qmc.py::TestUtils::test_geometric_discrepancy_mst_with_zero_distances
  /opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/scipy/stats/tests/test_qmc.py:233: UserWarning: Sample contains duplicate points.
    assert_allclose(qmc.geometric_discrepancy(sample, method='mst'), 0.5)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================= slowest 10 durations =============================
10.07s setup    _lib/tests/test_warnings.py::test_warning_calls_filters
9.99s call     stats/tests/test_continuous.py::TestDistributions::test_support_moments_sample[StandardNormal]
7.94s call     special/tests/test_cdflib.py::TestCDFlib::test_chndtrix
5.92s call     linalg/tests/test_extending.py::test_cython
5.26s call     stats/tests/test_mgc.py::TestMGCStat::test_pvalue_literature
4.22s call     stats/tests/test_multivariate.py::TestNormalInverseGamma::test_moments
3.98s call     _lib/tests/test_import_cycles.py::test_public_modules_importable
3.60s call     sparse/tests/test_sparsetools.py::TestInt32Overflow::test_bsr_n_block[matvec]
3.59s call     signal/tests/test_signaltools.py::TestConvolve2d::test_large_array
3.57s call     optimize/tests/test_optimize.py::TestOptimizeSimple::test_minimize_callback_copies_array[fmin]
=========================== short test summary info ============================
FAILED interpolate/tests/test_interpnd.py::TestLinearNDInterpolation::test_threading
= 1 failed, 72524 passed, 4586 skipped, 375 xfailed, 20 xpassed, 2591 warnings in 635.43s (0:10:35) =
```

mdhaber commented:
That is in main and reported in gh-21885.

rgommers committed:
Merge branch 'main' into newHighs

rgommers reviewed (approved):
All looks happy now, let's give it a go - fingers crossed:) Thanks a lot @HaoZeke!
